wrapper.gradle

```diff
@@ -17,9 +17,108 @@
  * under the License.
  */
 
-defaultTasks 'downloadWrapper'
+// This file contains tasks for the gradle wrapper generation.
 
-task downloadWrapper(type: Wrapper) {
-    description = "Download the gradle wrapper and requisite files. Overwrites existing wrapper files."
+// Ensure the wrapper script is generated based on the version defined in the project
+// and not the version installed on the machine running the task.
+// Read more about the wrapper here: https://docs.gradle.org/current/userguide/gradle_wrapper.html
+wrapper {
     gradleVersion = project.gradleVersion
-}
\ No newline at end of file
+    distributionType = Wrapper.DistributionType.ALL
+}
+
+def licenseString = """#
+# Copyright 2017 the original author or authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License."""
+
+// Custom task to inject support for downloading the gradle wrapper jar if it doesn't exist.
+// This allows us to avoid checking in the jar to our repository.
+// Additionally adds a license header to the wrapper while editing the file contents.
+task bootstrapWrapper() {
+    // In the doLast block so this runs when the task is called and not during project configuration.
+    doLast {
+        def wrapperBasePath = "\$APP_HOME/gradle/wrapper"
+        def wrapperJarPath = wrapperBasePath + "/gradle-wrapper.jar"
+
+        // Add a trailing zero to the version if needed.
+        def fullVersion = project.gradleVersion.count(".") == 1 ? "${project.gradleVersion}.0" : versions.gradle
```

Comment:
If I understand correctly here you would use the project's gradle version instead of the `distributionUrl` in the gradle-wrapper.properties, right? That would make sure to always use the correct version of gradle.
I think perhaps we could make a comment there that people should not expect the `distributionUrl` to work anymore, neither `gradle wrapper --gradle-version x.y` as they'd expect.

Reply:
All of those things would still work as expected and as it did before. 

What this patch does is edit the `gradlew` script to support downloading the wrapper jar if it doesn't exist. That jar is technically gradle version agnostic. It hasn't changed for a very long time. 

Running the `gradle wrapper` task will still output a `gradle/wrapper/gradle-wrapper.properties` file and passing `--gradle-version x.y.x`, `distribution-type=foo` or `--gradle-distribution-url=bar` will all still override the task defaults.